File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleDataFieldRow.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.sp
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** A label/value pair on the verification-details card, optionally with a copy control. */
@Composable
fun UseSmileIDSampleDataFieldRow(
    label: String,
    value: String,
    modifier: Modifier = Modifier,
    testId: String? = null,
    onCopy: (() -> Unit)? = null,
    copyTestId: String? = null,
    /** Set only where the design colours the value, as the Status row's HTTP code is. */
    valueColor: Color? = null,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .defaultMinSize(minHeight = SmileDimens.space40)
            .padding(horizontal = SmileDimens.spacingMd, vertical = SmileDimens.spacingSm)
            .tagged(testId),
        horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = label,
            style = UseSmileIDSampleTheme.type.dataFieldLabelFont.copy(fontSize = FIELD_TEXT_SIZE),
            color = UseSmileIDSampleTheme.colors.dataField.label,
        )
        // A column of its own, wrapping inside it: a FlowRow sent long values to a line below the label.
        Text(
            text = value,
            style = UseSmileIDSampleTheme.type.dataFieldValueFont.copy(
                fontSize = FIELD_TEXT_SIZE,
                fontWeight = if (valueColor == null) FontWeight.SemiBold else FontWeight.Bold,
            ),
            color = valueColor ?: UseSmileIDSampleTheme.colors.dataField.value,
            textAlign = TextAlign.End,
            modifier = Modifier.weight(1f),
        )
        if (onCopy != null) {
            CopyButton(label = label, onCopy = onCopy, testId = copyTestId)
        }
    }
}

@Composable
private fun CopyButton(label: String, onCopy: () -> Unit, testId: String?) {
    Surface(
        onClick = onCopy,
        modifier = Modifier
            .size(SmileDimens.sizeIconLg)
            .semantics { contentDescription = "Copy $label" }
            .tagged(testId),
        shape = RoundedCornerShape(SmileDimens.radiusSm),
        color = UseSmileIDSampleTheme.colors.surfaceTile,
    ) {
        Box(contentAlignment = Alignment.Center) {
            CopyGlyph(tint = UseSmileIDSampleTheme.colors.textMuted)
        }
    }
}

private val FIELD_TEXT_SIZE = 13.sp

File: flutter/sample_ui/lib/src/components/use_smileid_sample_data_field_row.dart
import 'package:flutter/material.dart';

import '../theme/use_smileid_sample_colors.dart';
import '../theme/use_smileid_sample_theme.dart';
import '../theme/use_smileid_sample_typography.dart';
import '../tokens/smile_tokens.dart';
import 'use_smileid_sample_glyphs.dart';

/// A label/value pair on the verification-details card, optionally with a copy control.
class UseSmileIDSampleDataFieldRow extends StatelessWidget {
  /// [valueColor] is set only where the design colours the value, as the Status row's code is.
  const UseSmileIDSampleDataFieldRow({
    required this.label,
    required this.value,
    this.onCopy,
    this.valueColor,
    this.testId,
    this.copyTestId,
    super.key,
  });

  /// The field name.
  final String label;

  /// The field value, already formatted — the data layer stores the code, not "200 OK".
  final String value;

  /// Copies the value; absent on rows the design gives no copy control.
  final VoidCallback? onCopy;

  /// Overrides the value's colour where the design colours it.
  final Color? valueColor;

  /// The `sample_*` id the screen supplies.
  final String? testId;

  /// The `sample_*` id for the copy control.
  final String? copyTestId;

  @override
  Widget build(BuildContext context) {
    final UseSmileIDSampleColors colors = UseSmileIDSampleTheme.colorsOf(
      context,
    );
    return Semantics(
      identifier: testId,
      child: ConstrainedBox(
        constraints: const BoxConstraints(minHeight: SmileDimens.space40),
        child: Padding(
          padding: const EdgeInsets.symmetric(
            horizontal: SmileDimens.spacingMd,
            vertical: SmileDimens.spacingSm,
          ),
          // The spec's wrapping rule, which the Compose twin does not yet follow.
          child: LayoutBuilder(
            builder: (BuildContext context, BoxConstraints constraints) => Wrap(
              spacing: SmileDimens.spacingXs,
              runSpacing: SmileDimens.spacingXxs,
              alignment: WrapAlignment.spaceBetween,
              crossAxisAlignment: WrapCrossAlignment.center,
              children: <Widget>[
                Text(
                  label,
                  style: UseSmileIDSampleType.dataFieldLabelFont.copyWith(
                    fontSize: _fieldTextSize,
                    color: colors.dataField.label,
                  ),
                ),
                Row(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    ConstrainedBox(
                      // The copy control's own width comes out of the value's column, or a value
                      // that fills the line pushes the control off the right edge.
                      constraints: BoxConstraints(
                        maxWidth: constraints.maxWidth - _copyReserve(onCopy),
                      ),
                      child: Text(
                        value,
                        textAlign: TextAlign.end,
                        style: UseSmileIDSampleType.dataFieldValueFont.copyWith(
                          fontSize: _fieldTextSize,
                          fontWeight: valueColor == null
                              ? FontWeight.w600
                              : FontWeight.w700,
                          color: valueColor ?? colors.dataField.value,
                        ),
                      ),
                    ),
                    if (onCopy != null) ...<Widget>[
                      const SizedBox(width: SmileDimens.spacingXs),
                      _CopyButton(
                        label: label,
                        onCopy: onCopy!,
                        testId: copyTestId,
                        colors: colors,
                      ),
                    ],
                  ],
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

class _CopyButton extends StatelessWidget {
  const _CopyButton({
    required this.label,
    required this.onCopy,
    required this.colors,
    this.testId,
  });

  final String label;
  final VoidCallback onCopy;
  final UseSmileIDSampleColors colors;
  final String? testId;

  @override
  Widget build(BuildContext context) => Semantics(
    identifier: testId,
    button: true,
    label: 'Copy $label',
    child: Material(
      color: colors.surfaceTile,
      borderRadius: BorderRadius.circular(SmileDimens.radiusSm),
      child: InkWell(
        onTap: onCopy,
        borderRadius: BorderRadius.circular(SmileDimens.radiusSm),
        child: SizedBox(
          width: SmileDimens.sizeIconLg,
          height: SmileDimens.sizeIconLg,
          child: Center(child: UseSmileIDSampleGlyphs.copy(colors.textMuted)),
        ),
      ),
    ),
  );
}

/// What the copy control occupies, so the value's column can leave room for it.
double _copyReserve(VoidCallback? onCopy) =>
    onCopy == null ? 0 : SmileDimens.sizeIconLg + SmileDimens.spacingXs;

/// Label and value alike; no type token carries 13.
const double _fieldTextSize = 13;
